Feature/dj rewrite

zwilias opened the pull request:
The goals are:

 - composable decoders
 - safely typed to the extent possible
 - well documented code
 - flexible (through `map/2`, `chain/2` and `succeed` and `fail`)
 - convenient
 - typed and accumulated errors

The API is quite far from legacy `dj`, but I feel this is a definite step
forward.

## TODO

- [x] Return `{error, [error(), ...]}` and scrap `{multiple, _}`
- [x] Document everything
- [x] Adapt readme
- [x] Pull tests apart
- [x] Improve coverage
- [x] Add `dj:uuid/1`

zwilias committed:
Ignore `doc` directory

zwilias committed:
Rewrite `dj`

The goals are:

 - composable decoders
 - safely typed to the extent possible
 - well documented code
 - flexible (through `map/2`, `chain/2` and `succeed` and `fail`)
 - convenient
 - typed and accumulated errors

The API is quite far from legacy `dj`, but I feel this is a definite step
forward.

zwilias committed:
Add guard to `mapn` and remove `bad_key` from test

By using `{labels, existing_atom}`, `bad_key` may generate an `invalid_json`
error.

The guard clause on `mapn` ensures that the provided function has the same arity
as the number of values that will be provided. Magic!

zwilias committed:
Docs

zwilias committed:
Move tests to separate file

zwilias committed:
Decoders return nonempty list of errors

zwilias committed:
Test coverage

zwilias committed:
Integrate `uuid`

zwilias commented on src/dj.erl:
```diff
@@ -1,222 +1,758 @@
+decode(Json, Decoder, Opts) ->
+  try
+    V = jsx:decode(Json, [return_maps | Opts]),
+    Decoder(V)
```
Perhaps this should be moved out of the try/catch so _only_ errors thrown by `jsx` become an `invalid_json` error.

sgillis replied:
I agree, would be nice if dj never fails since it doesn't do IO (unless you do custom maps of course)

zwilias committed:
Adapt readme, minimalistic style

zwilias committed:
Specify edoc options

zwilias committed:
Reorder functions to match exports and document types

zwilias committed:
Move specification of test command to `rebar.config`

Running `rebar3 check` locally is just less error prone than `rebar3 do eunit,
dialyzer, edoc` and hoping no one messes up.

zwilias committed:
Add elvis for Erlang style consistency

zwilias committed:
Bump jsx to 2.9.0 and add rebar.lock to git

zwilias committed:
Prevent elvis plugin from bleeding into consumers

Only adding the plugin in a specific profile prevents the plugin from being
installed when DJ is added as a consumer. As a sad side-effect, locally running
tests now becomes `rebar3 as lint check`.

Might have to add that to the README. Or write contribution guidelines.

zwilias committed:
Add `errors()` type for convenience

zwilias committed:
Expose `decoder(T)` to appease dialyzer

Dialyzer has a few outstanding bugs round success typing in combination with
opaque types. As a result, writing a `decoder(#{foo := bar})` - while correct
and working - generates a warning.

Exposing the internals, for whatever reason, fixes this.

zwilias committed:
Add `dj:set/1`

zwilias committed:
Add `nonempty_list/1`

sgillis commented on src/dj.erl:
```diff
@@ -1,222 +1,810 @@
+-spec atom() -> decoder(atom()).
+atom() ->
+  chain( value()
+       , fun (Json) when is_atom(Json)   -> succeed(Json);
```
You use `succeed` here, but not in the previous functions, why?

zwilias replied:
I sort of wanted to implement the convenience decoders with the public API of the module, built on top of the lower level functionality.

This is also reflected in the error messages, which end up as `{custom, {not_an_atom, V}}` rather than `{unexpected_type, bool_null_or_binary, V}`.

Perhaps I should more clearly section this off with a comment, to make it clear that the convenience decoders are only there for convenience, and don't require anything you couldn't do if `decoder` was an opaque type.

sgillis replied:
Ah, that makes sense

sgillis commented on src/dj.erl:
```diff
@@ -1,222 +1,810 @@
+  chain( binary()
+       , fun (V) ->
+             case re:run( V
+                        , <<"^[^@\s]+@([^.@\s]{2,}\.){1,}[a-z]{2,}$">>
```
I'm not an expert on email regexes, but how did you get to this one? I know there's some info on the subject at [emailregex.com](http://emailregex.com/)

zwilias replied:
I stole it from v1 and didn't really bother to look too closely at it. I'm definitely open to improving it, though 😅

zwilias replied:
After looking at this particular regex, it's actually not that horrible. It doesn't allow _everything_ that is technically valid (like having an IP address as the domain), but does a decent job at filtering out things that are "obviously invalid".

I'm gonna leave it for now, we can always circle back and upgrade this bit later on.

sgillis commented on src/dj.erl:
```diff
@@ -1,222 +1,810 @@
+    T   :: term(),
+    V   :: term().
+mapn(Fun, Decoders) when is_function(Fun, length(Decoders))->
+  map(fun (Vs) -> erlang:apply(Fun, Vs) end, sequence(Decoders)).
```
This will crash if the arity doesn't match right? Might be a more consistent if we return an error type in this case as well?

zwilias replied:
Right, good point. Agreed.

sgillis commented on src/dj.erl:
````diff
@@ -1,222 +1,810 @@
+%% ```
+%%    Dec = dj:to_map(#{ x => dj:index(0, dj:integer())
+%%                     , y => dj:index(1, dj:integer())
+%%                     , z => dj:index(2, dj:integer())
````
This might be premature optimisation, but this looks a bit inefficient since every `dj:index` will traverse the list

zwilias replied:
Yeah, it's a silly example. Using `dj:list` and `dj:chain` would definitely be a better idea for that particular use-case.

The main thing I wanted to demonstrate, though, is that `to_map/1` allows decoupling the Erlang representation from the JSON representation: it doesn't _have_ to be a JSON object.

Again, open to suggestions!

sgillis replied:
It gets the point across, so not really an issue probably

sgillis commented on src/dj.erl:
```diff
@@ -1,222 +1,810 @@
+%% with composition are discussed individually.
+-spec decode(Json, decoder(T), Opts) -> result(T, errors()) when
+    Json :: jsx:json_text(),
+    Opts :: [term()].
```
Why not create a nice typedef for the options as well? Makes it easier to see what the options are in my opinion

zwilias replied:
Yeah, I'm not actually sure what the options are! 😄 

(which indicates that it's a good idea, really)

zwilias commented on src/dj.erl:
```diff
@@ -1,222 +1,810 @@
+-spec decode(Json, decoder(T)) -> result(T, errors()) when
+    Json :: jsx:json_text().
+decode(Json, Decoder) ->
+  decode(Json, Decoder, [{labels, existing_atom}]).
```
@sgillis @kwrooijen I've been considering changing this to `attempt_atom`.

So the main reason for using `existing_atom` here is to minimize resource exhaustion attack surface, since atoms are never garbage collected.

However, this also means that `dj:decode/2` is incapable of decoding json that contains a key which is not a known atom.

By using `attempt_atom` instead, you could still _use_ `dj` as if it were using `existing_atom`: by only passing atoms to `dj:field/2` and `dj:at/2`, you create the atoms, hence ensuring they exist!

_If_ any unknown keys do exist, they'd simply stay binaries.

Wdyt? Are there downsides to such an approach?

kwrooijen replied:
IMHO, I think we should filter out fields that have not been specified in the DJ decoder. Why should we receive extra data we were not expecting?

zwilias replied:
To be fair, the only way that I can think of to actually access those fields is using `dj:value/0` which ostensibly returns the `jsx:json_term()` from the JSON.

Note that "data I don't need" doesn't necessarily mean "data I don't expect". When you have full control over the source of the JSON, you can indeed make sure it doesn't send any fields you're not interested in. When dealing with external API's however, you don't always have that control.

It's important here to remark that using decoders other than `dj:value/0` does filter out any data you don't need.

Are you suggesting to make `dj:decode/2` and `dj:decode/3` fail (or warn) when not all data in the json is actually used? ([prior art, in Elm](http://package.elm-lang.org/packages/zwilias/json-decode-exploration/5.0.0/Json-Decode-Exploration))

kwrooijen replied:
As discussed in private: I agree with `attempt_atom`

zwilias committed:
Switch to `attempt_atom`

`existing_atom` means we'd crash if the JSON contains any (unused) keys that
happen to not have a corresponding atom. `attempt_atom` means that all keys
referenced by `field/2` or `at/2` will become atoms, and unused fields are left
alone as binaries.

So:

 - no crashing
 - no resource exhaustion
 - same API

zwilias committed:
Move running decoders out of try/catch

zwilias committed:
Return error on arity mismatch for `dj:mapn/2`

zwilias committed:
Remove elvis plugin

It's outdated, requires some less-than-optimal hacks to run on travis, and fails
due to eunit macros.

zwilias committed:
Spec dj:decode/3 options
